# SQL Osnove

## Uvod

Ovaj notebook je moje kratko predstavljanje osnovnog znanja SQL-a. Temelji se na korištenju DUCKDB (https://duckdb.org/) database engine.

## Korišteni podaci

Ovi setovi podataka koriste u notebooku. Nije ih potrebno preuzimati jer im se može pristupiti direktno iz notebooka. 

- [gradovi.csv](https://raw.githubusercontent.com/Tomislav14/SQL-/main/cities%20(2)%20-%20Copy2.csv)
- [drzave.csv](https://raw.githubusercontent.com/Tomislav14/SQL-/main/countries%20(1).csv')



## Instalacija

Uncomment sljedeće ćelije kako bi instalirali tražene pakete. 

In [10]:
%pip install duckdb duckdb-engine jupysql

Note: you may need to restart the kernel to use updated packages.


## Importanje librarija i konfiguracija

In [11]:
import duckdb
import pandas as pd

# Importaj jupysql Jupyter ekstenzijuz kako bi kreirali SQL ćelije
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Postavi konfiguraciju na jupysql kako bi direktno i pojednostavljeno učitavali podatke na Pandas koji će ih potomo pročitatiu u Notebook. 

In [12]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

## Spajanje na DuckDB

Spoji jupysql na DuckDB kroz korištenje SQLAlchemy stila za konektiranje stringa. Na bazu se moguće spojiti u memoriju DuckDB ili file db. 

In [13]:
#%sql duckdb:///:memory:
%sql duckdb:///file.db

Ako je sql upit u jednom retku koda može koristiti  `%sql` magic naredba. Ako SQL upit ima više redova može se koristiti `%%sql` magic naredba.

## Instaliraj ekstenzije

Provjeri dostupne DuckDB ekstenzije

In [14]:
%%sql

SELECT * FROM duckdb_extensions();

extension_name  loaded  installed  \
0              arrow   False      False   
1       autocomplete   False      False   
2                aws   False      False   
3              azure   False      False   
4              excel   False      False   
5                fts   False       True   
6             httpfs    True       True   
7            iceberg   False      False   
8                icu    True       True   
9               inet   False      False   
10          jemalloc   False      False   
11              json    True       True   
12        motherduck   False      False   
13     mysql_scanner   False      False   
14           parquet    True       True   
15  postgres_scanner   False      False   
16           spatial   False       True   
17    sqlite_scanner   False      False   
18         substrait   False      False   
19             tpcds   False      False   
20              tpch    True       True   
21        visualizer   False      False   

                                         install_path  \
0                                                       
1                                                       
2                                                       
3                                                       
4                                                       
5                                          (BUILT-IN)   
6   C:\Users\tomi\.duckdb\extensions\v0.9.2\window...   
7                                                       
8                                          (BUILT-IN)   
9                                                       
10                                                      
11                                         (BUILT-IN)   
12                                                      
13                                                      
14                                         (BUILT-IN)   
15                                                      
16  C:\Users\tomi\.duckdb\extensions\v0.9.2\window...   
17                                                      
18                                                      
19                                                      
20                                         (BUILT-IN)   
21                                                      

                                          description            aliases  
0   A zero-copy data integration between Apache Ar...                 []  
1          Adds support for autocomplete in the shell                 []  
2        Provides features that depend on the AWS SDK                 []  
3   Adds a filesystem abstraction for Azure blob s...                 []  
4          Adds support for Excel-like format strings                 []  
5           Adds support for Full-Text Search Indexes                 []  
6   Adds support for reading and writing files ove...  [http, https, s3]  
7                     Adds support for Apache Iceberg                 []  
8   Adds support for time zones and collations usi...                 []  
9   Adds support for IP-related data types and fun...                 []  
10          Overwrites system allocator with JEMalloc                 []  
11                   Adds support for JSON operations                 []  
12     Enables motherduck integration with the system               [md]  
13    Adds support for connecting to a MySQL database            [mysql]  
14  Adds support for reading and writing parquet f...                 []  
15  Adds support for connecting to a Postgres data...         [postgres]  
16  Geospatial extension that adds support for wor...                 []  
17  Adds support for reading and writing SQLite da...  [sqlite, sqlite3]  
18         Adds support for the Substrait integration                 []  
19      Adds TPC-DS data generation and query support                 []  
20       Adds TPC-H data generation and query support                 []  
21  Creates an HTML-based visualization of the que... 

Duck DB omogućavaju vršenje upita na parquet i csv fileovima preko HTTP protokola (https://duckdb.org/docs/extensions/httpfs). To je korisno kod vršenja upita na velikim setovima podataka bez potrebe da se downloadaju na lokalnom računalu. 

In [15]:
%%sql

INSTALL httpfs;
LOAD httpfs;

Empty DataFrame
Columns: [Success]
Index: []

## RUčitaj CSV

Koristimo  `httpfs` ekstenziju kako bi pročitali  `cities.csv` file sa interneta.

In [16]:
%%sql

SELECT * FROM 'https://raw.githubusercontent.com/Tomislav14/SQL-/main/cities%20(2)%20-%20Copy2.csv';

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0        1           Bombo    UGA      0.58330     32.53330            75000
1        2     Fort Portal    UGA      0.67100     30.27500            42670
2        3         Potenza    ITA     40.64200     15.79900            69060
3        4      Campobasso    ITA     41.56300     14.65600            50762
4        5           Aosta    ITA     45.73700      7.31500            34062
...    ...             ...    ...          ...          ...              ...
1244  1245  Rio de Janeiro    BRA    -22.92502    -43.22502         11748000
1245  1246       Sao Paulo    BRA    -23.55868    -46.62502         18845000
1246  1247          Sydney    AUS    -33.92001    151.18518          4630000
1247  1248       Singapore    SGP      1.29303    103.85582          5183700
1248  1249       Hong Kong    CHN     22.30498    114.18501          7206000

[1249 rows x 6 columns]

In [17]:
%%sql

SELECT * FROM 'https://raw.githubusercontent.com/Tomislav14/SQL-/main/countries%20(1).csv';

id         ime_drzave alpha2_kod alpha3_kod  numericki_kod  geog_sirina  \
0      1        Afghanistan         AF        AFG              4      33.0000   
1      2            Albania         AL        ALB              8      41.0000   
2      3            Algeria         DZ        DZA             12      28.0000   
3      4     American Samoa         AS        ASM             16     -14.3333   
4      5            Andorra         AD        AND             20      42.5000   
..   ...                ...        ...        ...            ...          ...   
238  239  Wallis and Futuna         WF        WLF            876     -13.3000   
239  240     Western Sahara         EH        ESH            732      24.5000   
240  241              Yemen         YE        YEM            887      15.0000   
241  242             Zambia         ZM        ZMB            894     -15.0000   
242  243           Zimbabwe         ZW        ZWE            716     -20.0000   

     geog_duzina  
0           65.0  
1           20.0  
2            3.0  
3         -170.0  
4            1.6  
..           ...  
238       -176.2  
239        -13.0  
240         48.0  
241         30.0  
242         30.0  

[243 rows x 7 columns]

## Napravi tablicu

Napravi tablicu imena  `gradovi` iz `gradovi.csv` filea.

In [19]:
%%sql 

CREATE TABLE grad AS SELECT * FROM 'https://raw.githubusercontent.com/Tomislav14/SQL-/main/cities%20(2)%20-%20Copy2.csv';

Empty DataFrame
Columns: [Success]
Index: []

Napravi tablicu `drzave` iz `drzave.csv` filea.

In [21]:
%%sql 

CREATE TABLE drzava AS SELECT * FROM 'https://raw.githubusercontent.com/Tomislav14/SQL-/main/countries%20(1).csv';

Empty DataFrame
Columns: [Success]
Index: []

Prikaži tablicu u bazi podataka

In [22]:
%%sql 

FROM grad;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0        1           Bombo    UGA      0.58330     32.53330            75000
1        2     Fort Portal    UGA      0.67100     30.27500            42670
2        3         Potenza    ITA     40.64200     15.79900            69060
3        4      Campobasso    ITA     41.56300     14.65600            50762
4        5           Aosta    ITA     45.73700      7.31500            34062
...    ...             ...    ...          ...          ...              ...
1244  1245  Rio de Janeiro    BRA    -22.92502    -43.22502         11748000
1245  1246       Sao Paulo    BRA    -23.55868    -46.62502         18845000
1246  1247          Sydney    AUS    -33.92001    151.18518          4630000
1247  1248       Singapore    SGP      1.29303    103.85582          5183700
1248  1249       Hong Kong    CHN     22.30498    114.18501          7206000

[1249 rows x 6 columns]

In [23]:
%%sql 

FROM drzava;

id         ime_drzave alpha2_kod alpha3_kod  numericki_kod  geog_sirina  \
0      1        Afghanistan         AF        AFG              4      33.0000   
1      2            Albania         AL        ALB              8      41.0000   
2      3            Algeria         DZ        DZA             12      28.0000   
3      4     American Samoa         AS        ASM             16     -14.3333   
4      5            Andorra         AD        AND             20      42.5000   
..   ...                ...        ...        ...            ...          ...   
238  239  Wallis and Futuna         WF        WLF            876     -13.3000   
239  240     Western Sahara         EH        ESH            732      24.5000   
240  241              Yemen         YE        YEM            887      15.0000   
241  242             Zambia         ZM        ZMB            894     -15.0000   
242  243           Zimbabwe         ZW        ZWE            716     -20.0000   

     geog_duzina  
0           65.0  
1           20.0  
2            3.0  
3         -170.0  
4            1.6  
..           ...  
238       -176.2  
239        -13.0  
240         48.0  
241         30.0  
242         30.0  

[243 rows x 7 columns]

In [24]:
%%sql 

SELECT * FROM grad;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0        1           Bombo    UGA      0.58330     32.53330            75000
1        2     Fort Portal    UGA      0.67100     30.27500            42670
2        3         Potenza    ITA     40.64200     15.79900            69060
3        4      Campobasso    ITA     41.56300     14.65600            50762
4        5           Aosta    ITA     45.73700      7.31500            34062
...    ...             ...    ...          ...          ...              ...
1244  1245  Rio de Janeiro    BRA    -22.92502    -43.22502         11748000
1245  1246       Sao Paulo    BRA    -23.55868    -46.62502         18845000
1246  1247          Sydney    AUS    -33.92001    151.18518          4630000
1247  1248       Singapore    SGP      1.29303    103.85582          5183700
1248  1249       Hong Kong    CHN     22.30498    114.18501          7206000

[1249 rows x 6 columns]

In [25]:
%%sql

SELECT * FROM grad LIMIT 10;

id               ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0   1             Bombo    UGA      0.58330     32.53330            75000
1   2       Fort Portal    UGA      0.67100     30.27500            42670
2   3           Potenza    ITA     40.64200     15.79900            69060
3   4        Campobasso    ITA     41.56300     14.65600            50762
4   5             Aosta    ITA     45.73700      7.31500            34062
5   6         Mariehamn    ALD     60.09700     19.94900            10682
6   7          Ramallah    PSE     31.90294     35.20621            24599
7   8      Vatican City    VAT     41.90001     12.44781              832
8   9           Poitier    FRA     46.58329      0.33328            85960
9  10  Clermont-Ferrand    FRA     45.77998      3.08001           233050

In [26]:
%%sql

SELECT ime,drzava FROM grad LIMIT 10;

ime drzava
0             Bombo    UGA
1       Fort Portal    UGA
2           Potenza    ITA
3        Campobasso    ITA
4             Aosta    ITA
5         Mariehamn    ALD
6          Ramallah    PSE
7      Vatican City    VAT
8           Poitier    FRA
9  Clermont-Ferrand    FRA

In [27]:
%%sql

SELECT DISTINCT drzava FROM grad LIMIT 10;

drzava
0    PSE
1    VAT
2    NAM
3    NLD
4    GEO
5    PRT
6    PLW
7    MCO
8    PAK
9    KOR

In [28]:
%%sql

SELECT COUNT(*) FROM grad;

count_star()
0          1249

In [29]:
%%sql

SELECT COUNT(DISTINCT drzava) FROM grad;

count(DISTINCT drzava)
0                     200

In [30]:
%%sql

SELECT MAX(broj_stanovnika) FROM grad;

max(broj_stanovnika)
0              35676000

In [31]:
%%sql

SELECT SUM(broj_stanovnika) FROM grad;

sum(broj_stanovnika)
0          1.475535e+09

In [32]:
%%sql

SELECT AVG(broj_stanovnika) FROM grad;

avg(broj_stanovnika)
0          1.181373e+06

In [33]:
%%sql

SELECT * FROM grad ORDER BY drzava LIMIT 10;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0   666  Mazar-e Sharif    AFG     36.69999     67.10003           458151
1  1219           Kabul    AFG     34.51669     69.18326          3277000
2   665           Herat    AFG     34.33001     62.16999           481009
3  1024        Kandahar    AFG     31.61002     65.69495           715542
4  1177          Luanda    AGO     -8.83829     13.23443          5172900
5   669         Malanje    AGO     -9.54000     16.34003           125856
6   670        Benguela    AGO    -12.57826     13.40723           151226
7   671         Lubango    AGO    -14.91001     13.49002           125632
8   672          Namibe    AGO    -15.19004     12.16002           132900
9  1026        Menongue    AGO    -14.66661     17.69999            13030

In [34]:
%%sql 

SELECT * FROM grad ORDER BY drzava ASC, broj_stanovnika DESC LIMIT 10;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0  1219           Kabul    AFG     34.51669     69.18326          3277000
1  1024        Kandahar    AFG     31.61002     65.69495           715542
2   665           Herat    AFG     34.33001     62.16999           481009
3   666  Mazar-e Sharif    AFG     36.69999     67.10003           458151
4  1177          Luanda    AGO     -8.83829     13.23443          5172900
5  1027          Huambo    AGO    -12.74999     15.76001          1100000
6   670        Benguela    AGO    -12.57826     13.40723           151226
7   672          Namibe    AGO    -15.19004     12.16002           132900
8   669         Malanje    AGO     -9.54000     16.34003           125856
9   671         Lubango    AGO    -14.91001     13.49002           125632

In [35]:
%%sql

SELECT * FROM grad WHERE drzava='USA'

id                  ime drzava  geog_sirina  geog_duzina  \
0     115                Agana    USA     13.47002    144.75002   
1     124       San Bernardino    USA     34.12038   -117.30003   
2     125           Bridgeport    USA     41.17998    -73.19996   
3     126            Rochester    USA     43.17043    -77.61995   
4     160  International Falls    USA     48.60113    -93.41085   
..    ...                  ...    ...          ...          ...   
109  1186              Atlanta    USA     33.83001    -84.39995   
110  1187              Chicago    USA     41.82999    -87.75005   
111  1223          Los Angeles    USA     33.98998   -118.17998   
112  1224      Washington D.C.    USA     38.89955    -77.00942   
113  1225             New York    USA     40.74998    -73.98002   

     broj_stanovnika  
0             122411  
1            1745000  
2            1018000  
3             755000  
4              15240  
..               ...  
109          4506000  
110          8990000  
111         12500000  
112          4338000  
113         19040000  

[114 rows x 6 columns]

In [36]:
%%sql

SELECT * FROM grad WHERE drzava='USA' OR drzava='CAN';

id                  ime drzava  geog_sirina  geog_duzina  \
0     115                Agana    USA     13.47002    144.75002   
1     124       San Bernardino    USA     34.12038   -117.30003   
2     125           Bridgeport    USA     41.17998    -73.19996   
3     126            Rochester    USA     43.17043    -77.61995   
4     160  International Falls    USA     48.60113    -93.41085   
..    ...                  ...    ...          ...          ...   
154  1002        Charlottetown    CAN     46.24928    -63.13133   
155  1171             Edmonton    CAN     53.55002   -113.49998   
156  1172             Montreal    CAN     45.50000    -73.58330   
157  1216            Vancouver    CAN     49.27342   -123.12164   
158  1217              Toronto    CAN     43.69998    -79.42002   

     broj_stanovnika  
0             122411  
1            1745000  
2            1018000  
3             755000  
4              15240  
..               ...  
154            42402  
155          1058000  
156          3678000  
157          2313328  
158          5213000  

[159 rows x 6 columns]

In [37]:
%%sql 

SELECT * FROM grad WHERE drzava='USA' AND broj_stanovnika>1400000;

id              ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0    124   San Bernardino    USA     34.12038   -117.30003          1745000
1    183      Kansas City    USA     39.10709    -94.60409          1469000
2    188        Ft. Worth    USA     32.73998    -97.34004          1440454
3    202     Indianapolis    USA     39.74999    -86.17005          1436000
4    207        Baltimore    USA     39.29999    -76.61998          2255000
5    735         San Jose    USA     37.29998   -121.84999          1668000
6    736       Sacramento    USA     38.57502   -121.47004          1604000
7    737        Las Vegas    USA     36.21000   -115.22001          1823000
8    739         Portland    USA     45.52002   -122.67999          1875000
9    746      San Antonio    USA     29.48733    -98.50731          1473000
10   751        Cleveland    USA     41.46999    -81.69500          1890000
11   752       Cincinnati    USA     39.16188    -84.45692          1636000
12   758       Pittsburgh    USA     40.43000    -79.99999          1838000
13  1049         San Juan    USA     18.44002    -66.12998          2478905
14  1066      Minneapolis    USA     44.97998    -93.25179          2616000
15  1068          Seattle    USA     47.57000   -122.33998          3074000
16  1069          Phoenix    USA     33.53998   -112.06999          3551000
17  1070        San Diego    USA     32.82002   -117.17999          2916000
18  1071        St. Louis    USA     38.63502    -90.23998          2199000
19  1073           Dallas    USA     32.82002    -96.84002          4798000
20  1075           Boston    USA     42.32996    -71.07001          4467000
21  1076            Tampa    USA     27.94699    -82.45862          2314000
22  1077     Philadelphia    USA     39.99997    -75.17000          5492000
23  1078          Detroit    USA     42.32996    -83.08006          4101000
24  1182    San Francisco    USA     37.74001   -122.45998          3450000
25  1183           Denver    USA     39.73919   -104.98402          2313000
26  1184          Houston    USA     29.81997    -95.33998          4459000
27  1185            Miami    USA     25.78761    -80.22411          5585000
28  1186          Atlanta    USA     33.83001    -84.39995          4506000
29  1187          Chicago    USA     41.82999    -87.75005          8990000
30  1223      Los Angeles    USA     33.98998   -118.17998         12500000
31  1224  Washington D.C.    USA     38.89955    -77.00942          4338000
32  1225         New York    USA     40.74998    -73.98002         19040000

In [38]:
%%sql

SELECT * FROM grad WHERE drzava LIKE 'V%';

id               ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0      8      Vatican City    VAT     41.90001     12.44781              832
1    194            Merida    VEN      8.39999    -71.13001           345489
2    209   Puerto Ayacucho    VEN      5.66390    -67.62361            52526
3    229          Haiphong    VNM     20.83001    106.68009          1969000
4    230           Da Nang    VNM     16.06004    108.24997          1000000
5    715         Kingstown    VCT     13.14828    -61.21206            49485
6    747     San Cristobal    VEN      7.77000    -72.24997           438798
7    748          Valencia    VEN     10.22998    -67.98002          1770000
8    759    Ciudad Guayana    VEN      8.37002    -62.61999           746535
9    780         Port Vila    VUT    -17.73335    168.31664            44040
10  1074         Maracaibo    VEN     10.72998    -71.65998          2072000
11  1080             Hanoi    VNM     21.03333    105.85001          4378000
12  1081  Ho Chi Minh City    VNM     10.78003    106.69503          5314000
13  1188           Caracas    VEN     10.50100    -66.91704          2985000

In [39]:
%%sql

SELECT * FROM grad WHERE drzava LIKE '%N';

id        ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0      31       Gifu    JPN     35.42309    136.76275           412895
1     100  Matsuyama    JPN     33.84554    132.76584           533541
2     108  El Fasher    SDN     13.62998     25.35002           252609
3     122      Macau    CHN     22.20300    113.54505           568700
4     131     Xiamen    CHN     24.44999    118.08002          2519000
..    ...        ...    ...          ...          ...              ...
237  1235    Nairobi    KEN     -1.28335     36.81666          3010000
238  1236    Jakarta    IDN     -6.17442    106.82944          9125000
239  1239   Shanghai    CHN     31.21645    121.43650         14987000
240  1240      Tokyo    JPN     35.68502    139.75141         35676000
241  1249  Hong Kong    CHN     22.30498    114.18501          7206000

[242 rows x 6 columns]

In [40]:
%%sql 

SELECT * FROM grad WHERE drzava LIKE 'CR_';

id           ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0  449  Puerto Limon    CRI     10.00002    -83.03334            85001
1  902      San Jose    CRI      9.93501    -84.08405          1284000

In [41]:
%%sql

SELECT * FROM grad WHERE drzava IN ('JPN', 'AUS');

id            ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0     31           Gifu    JPN     35.42309    136.76275           412895
1    100      Matsuyama    JPN     33.84554    132.76584           533541
2    491        Fukuoka    JPN     33.59502    130.41001          2792000
3    492       Miyazaki    JPN     31.91824    131.41838           324384
4    493           Naha    JPN     26.20717    127.67297           905238
5    494          Kochi    JPN     33.56243    133.53752           335570
6    505         Nagoya    JPN     35.15500    136.91499          3230000
7    506         Nagano    JPN     36.65000    138.17001           594311
8    507        Kushiro    JPN     42.97496    144.37469           198566
9    508       Hakodate    JPN     41.79498    140.73998           302984
10   509          Kyoto    JPN     35.02999    135.75000          1805000
11   510         Sendai    JPN     38.28711    141.02172          2250000
12   511         Sakata    JPN     38.92004    139.85006           100446
13   921       Nagasaki    JPN     32.76499    129.88503           435455
14  1142      Hiroshima    JPN     34.38784    132.44291          2045000
15  1146        Sapporo    JPN     43.07498    141.34004          2544000
16  1208          Osaka    JPN     34.75004    135.46014         11294000
17  1240          Tokyo    JPN     35.68502    139.75141         35676000
18   684      Katherine    AUS    -14.46662    132.26659            10141
19   685      Busselton    AUS    -33.65641    115.34866            10214
20   686       Mandurah    AUS    -32.52348    115.74706            73356
21   687         Broome    AUS    -17.96177    122.23077            13218
22   688     Kalgoorlie    AUS    -30.73540    121.46002            36852
23   689         Albany    AUS    -35.01695    117.89160            26445
24   690   Port Hedland    AUS    -20.31040    118.60603            14288
25   691       Karratha    AUS    -20.73038    116.87001            16796
26   692      Geraldton    AUS    -28.76663    114.59997            27258
27   693       Griffith    AUS    -34.29001    146.04002            15455
28   694         Orange    AUS    -33.28000    149.09998            39329
29   695          Dubbo    AUS    -32.25996    148.59733            31574
30   696       Armidale    AUS    -30.51231    151.66748            22673
31   697    Broken Hill    AUS    -31.94995    141.43311            17395
32   698   Port Lincoln    AUS    -34.73324    135.86658            13630
33   699        Whyalla    AUS    -33.02503    137.56141            22559
34   700       Portland    AUS    -38.34000    141.59000            11808
35   701        Bendigo    AUS    -36.75999    144.28002            81657
36   702     Wangaratta    AUS    -36.36002    146.30000            14022
37   703       Windorah    AUS    -25.43324    142.65020              158
38   704      Mount Isa    AUS    -20.72387    139.49003            33200
39   705    Rockhampton    AUS    -23.36391    150.52000            65850
40   706         Cairns    AUS    -16.88784    145.76333           154225
41   707     Gold Coast    AUS    -28.08150    153.44825           527660
42   708      Devonport    AUS    -41.19267    146.33110            19317
43  1038         Darwin    AUS    -12.42535    130.85004            93080
44  1039  Alice Springs    AUS    -23.70100    133.88003            27710
45  1040       Canberra    AUS    -35.28303    149.12903           327700
46  1041      Newcastle    AUS    -32.84535    151.81501          1134616
47  1042       Adelaide    AUS    -34.93499    138.60000          1145000
48  1043     Townsville    AUS    -19.24995    146.77000           138954
49  1044       Brisbane    AUS    -27.45503    153.03509          1860000
50  1045         Hobart    AUS    -42.85001    147.29503            80870
51  1180          Perth    AUS    -31.95501    115.84000          1532000
52  1221      Melbourne    AUS    -37.82003    144.97502          4170000
53  1247  

In [42]:
%%sql 

SELECT * FROM grad WHERE broj_stanovnika BETWEEN 1700000 AND 10000000;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0     124  San Bernardino    USA     34.12038   -117.30003          1745000
1     127      Manchester    GBR     53.50042     -2.24799          2230000
2     129         Incheon    KOR     37.47615    126.64223          2550000
3     131          Xiamen    CHN     24.44999    118.08002          2519000
4     132        Nanchong    CHN     30.78043    106.13000          2174000
..    ...             ...    ...          ...          ...              ...
201  1242           Paris    FRA     48.86669      2.33334          9904000
202  1243        Santiago    CHL    -33.45001    -70.66704          5720000
203  1247          Sydney    AUS    -33.92001    151.18518          4630000
204  1248       Singapore    SGP      1.29303    103.85582          5183700
205  1249       Hong Kong    CHN     22.30498    114.18501          7206000

[206 rows x 6 columns]

## SQL Join



U ovom dijelu Notebooka napravljene su najčešće korištene vrste JOINA. 

- `(INNER) JOIN`: Vraća zapise koji se slažu iz obje tablice (Returns records that have matching values in both tables)
- `LEFT (OUTER) JOIN`: Vraća sve zapise iz lijeve tablice i zapise koji se slažu iz desne tablice (Returns all records from the left table, and the matched records from the right table)
- `RIGHT (OUTER) JOIN`: Vraća sve zapise iz desne tablice i zapise koji se slažu iz lijeve tablice (Returns all records from the right table, and the matched records from the left table)
- `FULL (OUTER) JOIN`: Vraća sve zapise iz obje tablice neovisno o tome je li se slažu (Returns all records when there is a match in either left or right table)



Imamo ukupno 1249 gradova u  `gradovi` tablici and 243 države u `države` tablici.

In [43]:
%%sql 

SELECT COUNT(*) FROM grad;

count_star()
0          1249

In [44]:
%%sql 

SELECT * FROM grad LIMIT 10;

id               ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0   1             Bombo    UGA      0.58330     32.53330            75000
1   2       Fort Portal    UGA      0.67100     30.27500            42670
2   3           Potenza    ITA     40.64200     15.79900            69060
3   4        Campobasso    ITA     41.56300     14.65600            50762
4   5             Aosta    ITA     45.73700      7.31500            34062
5   6         Mariehamn    ALD     60.09700     19.94900            10682
6   7          Ramallah    PSE     31.90294     35.20621            24599
7   8      Vatican City    VAT     41.90001     12.44781              832
8   9           Poitier    FRA     46.58329      0.33328            85960
9  10  Clermont-Ferrand    FRA     45.77998      3.08001           233050

In [45]:
%%sql 

SELECT COUNT(*) FROM drzava;

count_star()
0           243

In [46]:
%%sql 

SELECT * FROM drzava LIMIT 10;

id           ime_drzave alpha2_kod alpha3_kod  numericki_kod  geog_sirina  \
0   1          Afghanistan         AF        AFG              4      33.0000   
1   2              Albania         AL        ALB              8      41.0000   
2   3              Algeria         DZ        DZA             12      28.0000   
3   4       American Samoa         AS        ASM             16     -14.3333   
4   5              Andorra         AD        AND             20      42.5000   
5   6               Angola         AO        AGO             24     -12.5000   
6   7             Anguilla         AI        AIA            660      18.2500   
7   8           Antarctica         AQ        ATA             10     -90.0000   
8   9  Antigua and Barbuda         AG        ATG             28      17.0500   
9  10            Argentina         AR        ARG             32     -34.0000   

   geog_duzina  
0      65.0000  
1      20.0000  
2       3.0000  
3    -170.0000  
4       1.6000  
5      18.5000  
6     -63.1667  
7       0.0000  
8     -61.8000  
9     -64.0000

### SQL Inner Join

The `INNER JOIN` keyword selects records that have matching values in both tables. In the example, we join the `cities` table with the `countries` table using the `country` column in the `cities` table and the `Alpha3_code` column in the `countries` table. The result contains 1,244 rows, indicating that there are 5 cities that do not have a matching country.

In [49]:
%%sql

SELECT * FROM grad INNER JOIN drzava ON grad.drzava = drzava."alpha3_kod";

id          ime drzava  geog_sirina  geog_duzina  broj_stanovnika  \
0        1        Bombo    UGA      0.58330     32.53330            75000   
1        2  Fort Portal    UGA      0.67100     30.27500            42670   
2        3      Potenza    ITA     40.64200     15.79900            69060   
3        4   Campobasso    ITA     41.56300     14.65600            50762   
4        5        Aosta    ITA     45.73700      7.31500            34062   
...    ...          ...    ...          ...          ...              ...   
1239  1205       Berlin    DEU     52.52182     13.40155          3406000   
1240  1208        Osaka    JPN     34.75004    135.46014         11294000   
1241  1215       Tehran    IRN     35.67194     51.42434          7873000   
1242  1231  Mexico City    MEX     19.44244    -99.13099         19028000   
1243  1240        Tokyo    JPN     35.68502    139.75141         35676000   

      id_2                 ime_drzave alpha2_kod alpha3_kod  numericki_kod  \
0      226                     Uganda         UG        UGA            800   
1      226                     Uganda         UG        UGA            800   
2      106                      Italy         IT        ITA            380   
3      106                      Italy         IT        ITA            380   
4      106                      Italy         IT        ITA            380   
...    ...                        ...        ...        ...            ...   
1239    79                    Germany         DE        DEU            276   
1240   108                      Japan         JP        JPN            392   
1241   101  Iran, Islamic Republic of         IR        IRN            364   
1242   140                     Mexico         MX        MEX            484   
1243   108                      Japan         JP        JPN            392   

      geog_sirina_2  geog_duzina_2  
0            1.0000        32.0000  
1            1.0000        32.0000  
2           42.8333        12.8333  
3           42.8333        12.8333  
4           42.8333        12.8333  
...             ...            ...  
1239        51.0000         9.0000  
1240        36.0000       138.0000  
1241        32.0000        53.0000  
1242        23.0000      -102.0000  
1243        36.0000       138.0000  

[1244 rows x 13 columns]

Only select the `city` and `country` columns from the `cities` table and the `Country` column from the `countries` table.

In [50]:
%%sql

SELECT ime, grad."drzava", drzava."ime_drzave" FROM grad INNER JOIN drzava ON grad.drzava = drzava."alpha3_kod";

ime drzava                 ime_drzave
0           Bombo    UGA                     Uganda
1     Fort Portal    UGA                     Uganda
2         Potenza    ITA                      Italy
3      Campobasso    ITA                      Italy
4           Aosta    ITA                      Italy
...           ...    ...                        ...
1239       Berlin    DEU                    Germany
1240        Osaka    JPN                      Japan
1241       Tehran    IRN  Iran, Islamic Republic of
1242  Mexico City    MEX                     Mexico
1243        Tokyo    JPN                      Japan

[1244 rows x 3 columns]

### SQL Left Join

The `LEFT JOIN` keyword returns all records from the left table (`cities`), and the matched records from the right table (`countries`). The result contains 1,249 rows, the same number of rows as the `cities` table.

In [51]:
%%sql

SELECT * FROM grad LEFT JOIN drzava ON grad.drzava = drzava."alpha3_kod";

id          ime drzava  geog_sirina  geog_duzina  broj_stanovnika  \
0        1        Bombo    UGA      0.58330     32.53330            75000   
1        2  Fort Portal    UGA      0.67100     30.27500            42670   
2        3      Potenza    ITA     40.64200     15.79900            69060   
3        4   Campobasso    ITA     41.56300     14.65600            50762   
4        5        Aosta    ITA     45.73700      7.31500            34062   
...    ...          ...    ...          ...          ...              ...   
1244     6    Mariehamn    ALD     60.09700     19.94900            10682   
1245   118   Bir Lehlou    SAH     26.11917     -9.65252              500   
1246   415     Pristina    KOS     42.66671     21.16598           465186   
1247   719      Berbera    SOL     10.43555     45.01641           242344   
1248  1053     Hargeysa    SOL      9.56002     44.06531           477876   

       id_2 ime_drzave alpha2_kod alpha3_kod  numericki_kod  geog_sirina_2  \
0     226.0     Uganda         UG        UGA          800.0         1.0000   
1     226.0     Uganda         UG        UGA          800.0         1.0000   
2     106.0      Italy         IT        ITA          380.0        42.8333   
3     106.0      Italy         IT        ITA          380.0        42.8333   
4     106.0      Italy         IT        ITA          380.0        42.8333   
...     ...        ...        ...        ...            ...            ...   
1244    NaN       None       None       None            NaN            NaN   
1245    NaN       None       None       None            NaN            NaN   
1246    NaN       None       None       None            NaN            NaN   
1247    NaN       None       None       None            NaN            NaN   
1248    NaN       None       None       None            NaN            NaN   

      geog_duzina_2  
0           32.0000  
1           32.0000  
2           12.8333  
3           12.8333  
4           12.8333  
...             ...  
1244            NaN  
1245            NaN  
1246            NaN  
1247            NaN  
1248            NaN  

[1249 rows x 13 columns]

### SQL Right Join

The `RIGHT JOIN` keyword returns all records from the right table (`countries`), and the matched records from the left table (`cities`). The result contains 1,291 rows.

In [52]:
%%sql

SELECT * FROM grad RIGHT JOIN drzava ON grad.drzava = drzava."alpha3_kod";

id          ime drzava  geog_sirina  geog_duzina  broj_stanovnika  \
0     1.0        Bombo    UGA       0.5833      32.5333          75000.0   
1     2.0  Fort Portal    UGA       0.6710      30.2750          42670.0   
2     3.0      Potenza    ITA      40.6420      15.7990          69060.0   
3     4.0   Campobasso    ITA      41.5630      14.6560          50762.0   
4     5.0        Aosta    ITA      45.7370       7.3150          34062.0   
...   ...          ...    ...          ...          ...              ...   
1286  NaN         None   None          NaN          NaN              NaN   
1287  NaN         None   None          NaN          NaN              NaN   
1288  NaN         None   None          NaN          NaN              NaN   
1289  NaN         None   None          NaN          NaN              NaN   
1290  NaN         None   None          NaN          NaN              NaN   

      id_2                 ime_drzave alpha2_kod alpha3_kod  numericki_kod  \
0      226                     Uganda         UG        UGA            800   
1      226                     Uganda         UG        UGA            800   
2      106                      Italy         IT        ITA            380   
3      106                      Italy         IT        ITA            380   
4      106                      Italy         IT        ITA            380   
...    ...                        ...        ...        ...            ...   
1286   185  Saint Pierre and Miquelon         PM        SPM            666   
1287    85                 Guadeloupe         GP        GLP            312   
1288   162   Northern Mariana Islands         MP        MNP            580   
1289    83                  Greenland         GL        GRL            304   
1290   127                      Macao         MO        MAC            446   

      geog_sirina_2  geog_duzina_2  
0            1.0000        32.0000  
1            1.0000        32.0000  
2           42.8333        12.8333  
3           42.8333        12.8333  
4           42.8333        12.8333  
...             ...            ...  
1286        46.8333       -56.3333  
1287        16.2500       -61.5833  
1288        15.2000       145.7500  
1289        72.0000       -40.0000  
1290        22.1667       113.5500  

[1291 rows x 13 columns]

### SQL Full Join

The `FULL JOIN` keyword returns all records when there is a match in either left or right table. The result contains 1,296 rows.

In [53]:
%%sql

SELECT * FROM grad FULL JOIN drzava ON grad.drzava = drzava."alpha3_kod";

id          ime drzava  geog_sirina  geog_duzina  broj_stanovnika  \
0     1.0        Bombo    UGA       0.5833      32.5333          75000.0   
1     2.0  Fort Portal    UGA       0.6710      30.2750          42670.0   
2     3.0      Potenza    ITA      40.6420      15.7990          69060.0   
3     4.0   Campobasso    ITA      41.5630      14.6560          50762.0   
4     5.0        Aosta    ITA      45.7370       7.3150          34062.0   
...   ...          ...    ...          ...          ...              ...   
1291  NaN         None   None          NaN          NaN              NaN   
1292  NaN         None   None          NaN          NaN              NaN   
1293  NaN         None   None          NaN          NaN              NaN   
1294  NaN         None   None          NaN          NaN              NaN   
1295  NaN         None   None          NaN          NaN              NaN   

       id_2                 ime_drzave alpha2_kod alpha3_kod  numericki_kod  \
0     226.0                     Uganda         UG        UGA          800.0   
1     226.0                     Uganda         UG        UGA          800.0   
2     106.0                      Italy         IT        ITA          380.0   
3     106.0                      Italy         IT        ITA          380.0   
4     106.0                      Italy         IT        ITA          380.0   
...     ...                        ...        ...        ...            ...   
1291  185.0  Saint Pierre and Miquelon         PM        SPM          666.0   
1292   85.0                 Guadeloupe         GP        GLP          312.0   
1293  162.0   Northern Mariana Islands         MP        MNP          580.0   
1294   83.0                  Greenland         GL        GRL          304.0   
1295  127.0                      Macao         MO        MAC          446.0   

      geog_sirina_2  geog_duzina_2  
0            1.0000        32.0000  
1            1.0000        32.0000  
2           42.8333        12.8333  
3           42.8333        12.8333  
4           42.8333        12.8333  
...             ...            ...  
1291        46.8333       -56.3333  
1292        16.2500       -61.5833  
1293        15.2000       145.7500  
1294        72.0000       -40.0000  
1295        22.1667       113.5500  

[1296 rows x 13 columns]

### SQL Union

The `UNION` operator is used to combine the result-set of two or more `SELECT` statements.

In [54]:
%%sql

SELECT drzava FROM grad
UNION 
SELECT "alpha3_kod" FROM drzava;

drzava
0      PSE
1      VAT
2      NAM
3      NLD
4      GEO
..     ...
242    MNP
243    SHN
244    SJM
245    TCA
246    WLF

[247 rows x 1 columns]

In [55]:
%%sql

SELECT COUNT(ime), drzava
FROM grad 
GROUP BY drzava
ORDER BY COUNT(ime) DESC;

count(ime) drzava
0           114    USA
1           100    CHN
2            81    RUS
3            69    IND
4            46    BRA
..          ...    ...
195           1    LSO
196           1    EST
197           1    BGR
198           1    MDV
199           1    CYP

[200 rows x 2 columns]

In [57]:
%%sql

SELECT drzava."ime_drzave", COUNT(ime)
FROM grad
LEFT JOIN drzava ON grad.drzava = drzava."alpha3_kod"
GROUP BY drzava."ime_drzave"
ORDER BY COUNT(ime) DESC;

ime_drzave  count(ime)
0     United States         114
1             China         100
2            Russia          81
3             India          69
4            Brazil          46
..              ...         ...
192         Jamaica           1
193      Azerbaijan           1
194      Seychelles           1
195  Czech Republic           1
196           Samoa           1

[197 rows x 2 columns]

In [58]:
%%sql 

SELECT COUNT(ime), drzava
FROM grad
GROUP BY drzava
HAVING COUNT(ime) > 40
ORDER BY COUNT(ime) DESC;

count(ime) drzava
0         114    USA
1         100    CHN
2          81    RUS
3          69    IND
4          46    BRA
5          45    CAN

In [60]:
%%sql

SELECT drzava."ime_drzave", COUNT(ime)
FROM grad
LEFT JOIN drzava ON grad.drzava = drzava."alpha3_kod"
GROUP BY drzava."ime_drzave"
HAVING COUNT(ime) > 40
ORDER BY COUNT(ime) DESC;

ime_drzave  count(ime)
0  United States         114
1          China         100
2         Russia          81
3          India          69
4         Brazil          46
5         Canada          45

In [61]:
%%sql

SELECT ime, broj_stanovnika,
CASE
    WHEN broj_stanovnika> 10000000 THEN 'Megacity'
    WHEN broj_stanovnika > 1000000 THEN 'Large city'
    ELSE 'Small city'
END AS category
FROM grad;

ime  broj_stanovnika    category
0              Bombo            75000  Small city
1        Fort Portal            42670  Small city
2            Potenza            69060  Small city
3         Campobasso            50762  Small city
4              Aosta            34062  Small city
...              ...              ...         ...
1244  Rio de Janeiro         11748000    Megacity
1245       Sao Paulo         18845000    Megacity
1246          Sydney          4630000  Large city
1247       Singapore          5183700  Large city
1248       Hong Kong          7206000  Large city

[1249 rows x 3 columns]

## Saving results

You can save the results of a query to a new table using the `CREATE TABLE AS` statement.

In [62]:
%%sql

CREATE TABLE gradovi3 AS SELECT * FROM grad;

Empty DataFrame
Columns: [Success]
Index: []

Show the new table content.

In [63]:
%%sql

FROM gradovi3;

id             ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0        1           Bombo    UGA      0.58330     32.53330            75000
1        2     Fort Portal    UGA      0.67100     30.27500            42670
2        3         Potenza    ITA     40.64200     15.79900            69060
3        4      Campobasso    ITA     41.56300     14.65600            50762
4        5           Aosta    ITA     45.73700      7.31500            34062
...    ...             ...    ...          ...          ...              ...
1244  1245  Rio de Janeiro    BRA    -22.92502    -43.22502         11748000
1245  1246       Sao Paulo    BRA    -23.55868    -46.62502         18845000
1246  1247          Sydney    AUS    -33.92001    151.18518          4630000
1247  1248       Singapore    SGP      1.29303    103.85582          5183700
1248  1249       Hong Kong    CHN     22.30498    114.18501          7206000

[1249 rows x 6 columns]

Use the `DROP TABLE` statement to delete the table.

In [64]:
%%sql

DROP TABLE IF EXISTS gradovi_sad;
CREATE TABLE gradovi_sad AS (SELECT * FROM grad WHERE drzava = 'USA');

Empty DataFrame
Columns: [Success]
Index: []

In [65]:
%%sql

FROM gradovi_sad;

id                  ime drzava  geog_sirina  geog_duzina  \
0     115                Agana    USA     13.47002    144.75002   
1     124       San Bernardino    USA     34.12038   -117.30003   
2     125           Bridgeport    USA     41.17998    -73.19996   
3     126            Rochester    USA     43.17043    -77.61995   
4     160  International Falls    USA     48.60113    -93.41085   
..    ...                  ...    ...          ...          ...   
109  1186              Atlanta    USA     33.83001    -84.39995   
110  1187              Chicago    USA     41.82999    -87.75005   
111  1223          Los Angeles    USA     33.98998   -118.17998   
112  1224      Washington D.C.    USA     38.89955    -77.00942   
113  1225             New York    USA     40.74998    -73.98002   

     broj_stanovnika  
0             122411  
1            1745000  
2            1018000  
3             755000  
4              15240  
..               ...  
109          4506000  
110          8990000  
111         12500000  
112          4338000  
113         19040000  

[114 rows x 6 columns]

Use the `INSERT INTO` statement to insert rows into a table.

In [66]:
%%sql 

INSERT INTO gradovi_sad (SELECT * FROM grad WHERE drzava = 'CAN');

Empty DataFrame
Columns: [Success]
Index: []

In [ ]:
%%sql

SELECT * FROM gradovi LIMIT 10 -- Ovo je komentar;

id               ime drzava  geog_sirina  geog_duzina  broj_stanovnika
0   1             Bombo    UGA      0.58330     32.53330            75000
1   2       Fort Portal    UGA      0.67100     30.27500            42670
2   3           Potenza    ITA     40.64200     15.79900            69060
3   4        Campobasso    ITA     41.56300     14.65600            50762
4   5             Aosta    ITA     45.73700      7.31500            34062
5   6         Mariehamn    ALD     60.09700     19.94900            10682
6   7          Ramallah    PSE     31.90294     35.20621            24599
7   8      Vatican City    VAT     41.90001     12.44781              832
8   9           Poitier    FRA     46.58329      0.33328            85960
9  10  Clermont-Ferrand    FRA     45.77998      3.08001           233050

In [ ]:
%%sql

SELECT COUNT(ime), drzava
FROM gradovi 
/*
 * Dodaj Group by
 * Dodaj Order by
 */
GROUP BY drzava
ORDER BY COUNT(ime) DESC
LIMIT 10;

count(ime) drzava
0         114    USA
1         100    CHN
2          81    RUS
3          69    IND
4          46    BRA
5          45    CAN
6          40    ATA
7          36    AUS
8          30    FRA
9          27    MEX